File: GeeksforGeeks_Solutions/C++/October_13_25.cpp
class solution{
    public:
    unordered_map<Node*,int> mymap;
   
   int getMaxSum(Node *root) 
   {
       // Add your code here
       if(!root) return 0;
       
       if(mymap.find(root)!=mymap.end()) return mymap[root];
       
       int inc = root->data;
       if(root->left){
           inc += getMaxSum(root->left->left);
           inc += getMaxSum(root->left->right);
       }
       if(root->right){
           inc+=getMaxSum(root->right->left);
           inc+=getMaxSum(root->right->right);
       }
       
       int exc = getMaxSum(root->right)+getMaxSum(root->left);
       
       
       mymap[root] = max(inc,exc);
       
       return mymap[root];
   }
};

File: GeeksforGeeks_Solutions/Java/october_13_25.java
class Solution
{
    static HashMap<Node,Integer> hashMap =new HashMap<>();
    static int getMaxSum(Node root)
    {
        if(root==null){
            return 0;
        }
        if(hashMap.containsKey(root)){
            return hashMap.get(root);
        }
        int counter = root.data;
        if(root.left!=null){
            counter += getMaxSum(root.left.left);
            counter += getMaxSum(root.left.right);
        }
        if(root.right!=null){
            counter += getMaxSum(root.right.left);
            counter += getMaxSum(root.right.right);
        }
        int out = getMaxSum(root.left)+getMaxSum(root.right);
        hashMap.put(root,Math.max(counter, out));
        return hashMap.get(root);
    }
}